### ISSUE TITLE
Javascript Classes...when using multiple calls to jerry_eval

### ISSUE THREAD

**Issue by Dave-Osborne** [jerryscript-project/jerryscript#4839]
Hi,

I've just started using JerryScript, it looks great! I have a question around Javascript classes though...

The following executes as expected when running it with jerry_eval...

```
class Foo {
   bar() {
     print('Hello World!');
   }
};

var foo = new Foo();
foo.bar();
```

However, when calling jerry_eval to create the class......

```
class Foo {
   bar() {
     print('Hello World!');
   }
};
```

...and then calling jerry_eval to use the class fails...

```
var foo = new Foo();
foo.bar();
```

Perhaps this is a scope thing? or deliberately designed this way to minimise memory usage?

Thanks in advance!

**Comment by dbatyai** [comment]
Hi Dave!

This is indeed scoping related, there are two key parts to this.

The first thing is that classes are lexically scoped. This is similar to how variables declared with `let` work, they can only be referenced until the lexical environment they were declared in goes out of scope.

```js
class Foo {};

{
  class Bar {};
  
  // Both Foo and Bar are usable here
}

// Foo is usable, Bar will throw a ReferenceError
```

Secondly, eval calls are always evaluated in a new lexical environment, which is destroyed once the eval call is finished.
In the first example, both the class declaration and constructor call are in the same lexical environment that was created by the eval call, and so `Foo` can be resolved.

In the second example however, the lexical environment where the class was declared will go out of scope once the first eval call is finished. The second eval will be unable to reference any lexical variables from the previous lexical environment. It would roughly be equivalent to the following code:

```js
// First eval
{
  class Foo {};
}

// Second eval
{
  new Foo();
}
```

**Comment by rerobika** [comment]
According to the detailed description above the original issue seems resolved to me.
